libcontainer/factory_linux.go

```diff
@@ -261,19 +265,26 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 			fmt.Fprintln(os.Stderr, err)
 			return
 		}
-		// ensure that this pipe is always closed
-		pipe.Close()
 	}()
 	defer func() {
 		if e := recover(); e != nil {
 			err = fmt.Errorf("panic from initialization: %v, %v", e, string(debug.Stack()))
 		}
 	}()
+
+	var i initer
 	i, err = newContainerInit(it, pipe, rootfd)
 	if err != nil {
 		return err
 	}
-	return i.Init()
+
+	if i, err = newContainerInit(it, pipe, rootfd); err != nil {
+		return
```

Comment:
> You can't use i, err := as there are no new variables so it would fail compile.

Sorry, I missed that.

> Is there a reason for not using straight return as this is not only the idiomatic way

Because it's confusing to read.

> given the use of err in the defer not doing so could easily lead to additional coding errors where the defer fails to get the required error.

That will not happen with `return err`. However, it _can_ happen with the bare `return` format because it's less clear what the last value of `err` was. Then someone "refactors" the code and now you're forgetting to set an error.

Bare `return`s are one of the larger warts in Go IMO.